File: Algorithms/Sorting/Merge_Sort.cpp
//Name : Shivam Singhal
//Email ID : shivamsinghal0610@gmail.com

#include <bits/stdc++.h>
using namespace std;

void merge(long arr[], long l, long m, long r){
    long i, j, k;  
    //number of elements in first array (left half)
    long n1 = m - l + 1; 
    //number of elements in second array (right half)
    long n2 =  r - m; 
    //temporary arrays to store sorted sub-arrays 
    long L[n1], R[n2]; 
    for (i = 0; i < n1; i++)
        L[i] = arr[l + i];
    for (j = 0; j < n2; j++)
        R[j] = arr[m + 1+ j];
    //Initial index of first subarray
    i = 0; 
    //Initial index of second subarray
    j = 0; 
    //Initial index of merged subarray in original array
    k = l; 
    while (i < n1 && j < n2){
        //picking element from first sub-array
        if (L[i] <= R[j]){ 
            arr[k++] = L[i++];
        }
        //picking element from second sub-array
        else{ 
            arr[k++] = R[j++];
        }
    }
    //picking remaining elements of left sub-array, if any are left
    while (i < n1){ 
        arr[k++] = L[i++];
    }
    //picking remaining elements of right sub-array, if any are left
    while (j < n2){  
        arr[k++] = R[j++];
    }
}
 
void mergeSort(long arr[], long low, long high){
    if (low < high){
        //using this concept to avoid overflow in case of large numbers
        long mid = low + ((high-low)/2); 
        //sorting left half of array
        mergeSort(arr, low, mid); 
        //sorting right half of array
        mergeSort(arr, mid+1, high);
        //merging both sorted sub-arrays in a sorted fashion 
        merge(arr, low, mid, high); 
    }
}

void printArray(long a[], long n){
    for(long i=0; i<n ;i++)
     printf("%lld ",a[i]);
    printf("\n");
}
 
int main()
{
    long n;
    printf("Please enter the number of elements."); 
    scanf("%lld",&n);
    long arr[n];
    printf("\nPlease enter elements to be sorted.\n");
    for(long i=0; i<n ;i++)
     scanf("%lld",&arr[i]);
    printf("Entered elements :\n");
    printArray(arr,n);
    //function to sort array in ascending order using merge sort
    mergeSort(arr, 0,n-1); 
    printf("\nSorted array:\n");
    printArray(arr,n);
    return 0;
}

/*
Time Complexity : O(nlogn). 
The entire input is iterated over. This occurs O(log(n)) times (the input can only be halved O(log(n)) times). n items iterated log(n) times gives O(n log(n)).

Auxiliary Space Complexity : O(n). We use an extra array of n size. 
*/

File: Algorithms/Sorting/mergesort.java
import java.util.Scanner;

public class mergsort {
	@SuppressWarnings("resource")
	public static int[] takeInput() {
		Scanner s = new Scanner(System.in);
		int size = s.nextInt();
		int arr[] = new int[size];
		for (int i = 0; i < size; i++) {
			arr[i] = s.nextInt();
		}
		return arr;
	}
	
	public static void printArray(int input[]) {
		for(int i = 0; i < input.length; i++) {
			System.out.print(input[i] + " ");
		}
	}
	
	public static void main(String[] args) {
		int[] input = takeInput();
		mergeSort(input);
		printArray(input);
	}
/*********************************************/
public static void merge(int[] input,int s,int m,int e){  
         int i,j;
         int l=m-s+1;
         int p=e-m;
         int arr[]=new int[l];
         int arr2[]=new int [p];
        
        for(i=0;i<l;i++){
            arr[i]=input[s+i];
        }
        for(i=0;i<p;i++){
            arr2[i]=input[m+1+i];
        }
        i=0;
        j=0;
        int k=s;
        while(i<l&&j<p){
             if(arr[i]<=arr2[j]){
                input[k]=arr[i];
                i++;
             }
           else{
               input[k]=arr2[j];
               j++;
        }k++;
    }
         while (i < l) 
    { 
        input[k] = arr[i]; 
        i++; 
        k++; 
    } 
  
       while (j < p) 
    { 
        input[k]=arr2[j]; 
        j++; 
        k++; 
    } 
        
} 
            
   public static void sort(int[] input,int s,int e){
		if(s>=e){
            return ;
         }   
             int m=(s+e)/2;
               sort(input,s,m);
               sort(input,m+1,e);
               merge(input,s,m,e);
	}

    
	public static void mergeSort(int[] input){
		   sort(input , 0 ,input.length-1);
         }
}
